### PR TITLE
[RFR] Pass handleSubmit directly to buttons

### PR DESCRIPTION
**Pull request by djhi**
> No breaking changes, `handleSubmitWithRedirect` is still passed

## Use case

Having a custom save button which alter the values before actually saving them (see the `Save as Commentable` button in the `PostCreate` component of the `simple` example)

## Current issue

It's currently not possible to alter the form values before dispatching the save action:
- Dispatching a `redux-form` `change` action is asynchronous so we can't submit directly after changing the value
- Custom saga doesn't work either as it is not possible to dispatch a `redux-form` `submit` action because we don't have an `onSubmit` on our forms

## Solution

Passing the `redux-form` `handleSubmit` function to the buttons allows the user to do whatever he wants and still benefits from the default validation mechanisms

## TODO

- [x] Implementation
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by fzaninotto** [comment]
What about when the user submits the form by pressing `<Enter>`? The custom handler won't be called...

File: cypress/integration/create.js [hunk 1]
```diff
@@ -51,11 +51,38 @@ describe('Create Page', () => {
...
-            .should(el => expect(el).to.have.value('')); // new empty form
+        cy.get(CreatePage.elements.input('title')).should(el =>
+            expect(el).to.have.value('')
+        ); // new empty form
 
         ShowPage.navigate();
+        ShowPage.ShowPage.delete();
+    });
+
+    it.only('should allow to call a custom action updating values before submit', () => {
```
Comment (fzaninotto) [inline cypress/integration/create.js]: .only spotted

File: packages/ra-ui-materialui/src/input/BooleanInput.js [hunk 2]
```diff
@@ -24,14 +24,17 @@ export class BooleanInput extends Component {
             ...rest
         } = this.props;
 
+        const { value, ...inputProps } = input;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/BooleanInput.js]: this change seems unrelated, what's the motivation?

File: examples/simple/src/posts/PostCreate.js [hunk 3]
```diff
@@ -13,6 +16,44 @@ import {
     Toolbar,
 } from 'react-admin'; // eslint-disable-line import/no-unresolved
 
+const saveWithNoteButton = (values, basePath, redirectTo) =>
+    crudCreate('posts', { ...values, average_note: 10 }, basePath, redirectTo);
+
+const SaveWithNoteButton = connect(
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostCreate.js]: I don't like the direct HOC decoration, can you add an intermediary component?

File: examples/simple/src/posts/PostCreate.js [hunk 4]
```diff
@@ -13,6 +16,44 @@ import {
     Toolbar,
 } from 'react-admin'; // eslint-disable-line import/no-unresolved
 
+const saveWithNoteButton = (values, basePath, redirectTo) =>
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostCreate.js]: would be better named `saveWithNote`, as it's independent of the button context

File: packages/ra-ui-materialui/src/input/BooleanInput.js (hunk 2 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/BooleanInput.js]: Can be done in another PR. I stumbled upon this one while testing this feature. `BooleanInput` do not pass the input props (`name`, etc.) to the inner input element

Comment (djhi) [reply packages/ra-ui-materialui/src/input/BooleanInput.js]: I found this while trying to set a `BooleanInput` value with Cypress.

Comment (djhi) [reply packages/ra-ui-materialui/src/input/BooleanInput.js]: Should I open a PR on master for this ?

File: docs/Actions.md [hunk 5]
```diff
@@ -455,6 +455,69 @@ And that's all it takes to make a fetch action optimistic. Note that the `startU
...
 The fact that react-admin updates the internal store if you use custom actions with the `fetch` meta should be another motivation to avoid using raw `fetch`.
 
+## Altering the Form Values before Submitting
+
+Sometimes, you may want your custom action to alter the form values before actually sending them to the `dataProvider`. For those cases, you should know that every buttons inside a form [Toolbar](/CreateEdit.html#toolbar) receive two props:
+
+- `handleSubmitWithRedirect` which calls the default form save methods
+- `handleSubmit` which is the same prop as in [`react-form`](https://redux-form.com/7.4.2/docs/api/props.md/#-code-handlesubmit-eventorsubmit-function-code-)
+
+Knowing this, we can dispatch a custom action with a button. For instance, in the `simple` example:
```
Comment (fzaninotto) [inline docs/Actions.md]: s/ we can dispatch/ you can dispatch/

File: docs/Actions.md [hunk 6]
```diff
@@ -455,6 +455,69 @@ And that's all it takes to make a fetch action optimistic. Note that the `startU
...
+
+```jsx
+// A custom action creator which modifies the values before calling the default crudCreate action creator
+const saveWithNote = (values, basePath, redirectTo) =>
```
Comment (fzaninotto) [inline docs/Actions.md]: please add the `import { crudCreate } from 'react-admin'` to make your example easier to copy and paste.

File: docs/Actions.md [hunk 7]
```diff
@@ -455,6 +455,69 @@ And that's all it takes to make a fetch action optimistic. Note that the `startU
...
+            label="post.action.save_with_average_note"
+            redirect="show"
+            submitOnEnter={false}
+            variant="flat"
+        />
+    </Toolbar>
+);
+```
+
+The advantages of this is that we still benefit from the default `crudCreate` action side effects.
```
Comment (fzaninotto) [inline docs/Actions.md]: The advantages over what? What are the default action side effects? Also, prefer 'you' over 'we' in the documentation.

File: docs/CreateEdit.md [hunk 8]
```diff
@@ -630,12 +630,69 @@ Here are the props received by the `Toolbar` component when passed as the `toolb
...
 * `pristine`: A boolean indicating whether the form is pristine (eg: no inputs have been changed yet)
 * `redirect`: The default form's redirect
 * `saving`: A boolean indicating whether a save operation is ongoing.
 * `submitOnEnter`: A boolean indicating whether the form should be submitted when pressing `enter`
 
+**Tip**: To alter the form values before submitting, you should use the `handleSubmit` prop. For instance, in the `simple` example:
+
+```jsx
+// A custom action creator which modifies the values before calling the default crudCreate action creator
+const saveWithNote = (values, basePath, redirectTo) =>
```
Comment (fzaninotto) [inline docs/CreateEdit.md]: Please link to the other page rather than duplicating the example. You're doubling the cost of documentation maintenance.

**Comment by djhi** [comment]
> I still don't know what happens when the user submits the form with <Enter>. Nothing changed on that side?

Yup nothing changed :)

**Comment by djhi** [comment]
> What about when the user submits the form by pressing `<Enter>`? The custom handler won't be called...

Just tested it locally, if the button with the custom handler is the only one with a `submitOnEnter` set to `true` (you can only have one anyway), it works as expected

File: docs/Actions.md [hunk 9]
```diff
@@ -455,6 +455,67 @@ And that's all it takes to make a fetch action optimistic. Note that the `startU
...
+
+Sometimes, you may want your custom action to alter the form values before actually sending them to the `dataProvider`. For those cases, you should know that every buttons inside a form [Toolbar](/CreateEdit.html#toolbar) receive two props:
+
+- `handleSubmitWithRedirect` which calls the default form save methods
+- `handleSubmit` which is the same prop as in [`react-form`](https://redux-form.com/7.4.2/docs/api/props.md/#-code-handlesubmit-eventorsubmit-function-code-)
+
+Knowing this, you can dispatch a custom action with a button and still benefit from the default crud action side effects (notifications, optimistic ui, undo, etc.). For instance, in the `simple` example:
+
+```jsx
+// A custom action creator which modifies the values before calling the default crudCreate action creator
```
Comment (fzaninotto) [inline docs/Actions.md]: please add the `import { crudCreate } from 'react-admin'` line to make your example easier to copy and paste.

File: packages/ra-ui-materialui/src/input/BooleanInput.js (hunk 2 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/input/BooleanInput.js]: nevermind, let's keep it here
